### PR TITLE
Knockback 1.0

### PR DESCRIPTION
**Pull request by kmalakoff**
With the release of Knockback 1.0, I have updated the Knockback example and have ran the tests to make sure it functionally complies...everything passed.

### CODE REVIEW

**Comment by samccone** [comment]
@kmalakoff looks like a few jscs errors on your PR, would you mind addressing those?

thanks!

**Comment by kmalakoff** [comment]
It looks like these are style compliance problems due to CoffeeScript to JavaScript compilation.

I have tried to follow the style guidelines, but to adapt them to CoffeeScript.

What do you suggest as a solution since I do not have control over the generated JavaScript code?

**Comment by samccone** [comment]
hmm imho not using coffeescript would be the best fix :) Let me know what you think heh

**Comment by kmalakoff** [comment]
What I think...I'm not a fan of JavaScript and skipped learning it in any depth by working in CoffeeScript. ES6/ES7 are making JavaScript bearable.

Do any projects use "compile to" languages and how do you handle them? 
- If I'm one of the only holdouts from early TodoMVC days, I can agree to going through the generated JavaScript by hand and try to make it comply.
- if there are other holdouts, I'd rather save my time and have this waived

**Comment by samccone** [comment]
There are two other examples that are written in coffeescript https://github.com/tastejs/todomvc/search?utf8=%E2%9C%93&q=coffeescript&type=Code

That said, I am ok with ignoring these compile targets right now, and rewriting to JS as another issue/pr. How does this sound? :palm_tree:

**Comment by kmalakoff** [comment]
Thank you for the offer of an exception. Sounds good.

Based on your link, I see the policy is "Apps should be written without any preprocessors (Sass/CoffeeScript/..) to reach the largest audience"

Since it is only Serenadejs, Knockback, and Spine remaining, I'm fine with re-writing in POJS to help with the comparability. I'll do this and resubmit.

**Comment by kmalakoff** [comment]
@samccone I spent a couple of hours and updated the code to Plain Old JavaScript.

All of the tests pass fine for me locally, but some failed on Travis. Any idea about why there would be a difference?

```
  TodoMVC - knockback
    When page is initially opened
      ✓ should focus on the todo input field (39ms)
    No Todos
      ✓ should hide #main and #footer (194ms)
    New Todo
      ✓ should allow me to add todo items (1016ms)
      ✓ should clear text input field when an item is added (442ms)
      ✓ should append new items to the bottom of the list (980ms)
      ✓ should trim text input (406ms)
      ✓ should show #main and #footer when items added (475ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (310ms)
      ✓ should allow me to clear the completion state of all items (407ms)
      ✓ complete all checkbox should update state when items are completed / cleared (564ms)
    Item
      ✓ should allow me to mark items as complete (1029ms)
      ✓ should allow me to un-mark items as complete (943ms)
    Editing
      ✓ should hide other controls when editing (70ms)
      ✓ should save edits on enter (805ms)
      ✓ should save edits on blur (1045ms)
      ✓ should trim entered text (848ms)
      ✓ should remove the item if an empty text string was entered (657ms)
      ✓ should cancel edits on escape (762ms)
    Counter
      ✓ should display the current number of todo items (616ms)
    Clear completed button
      ✓ should display the correct text (186ms)
      ✓ should remove completed items when clicked (566ms)
      ✓ should be hidden when there are no items that are completed (510ms)
    Persistence
      ✓ should persist its data (1887ms)
    Routing
      ✓ should allow me to display active items (644ms)
      ✓ should respect the back button (1659ms)
      ✓ should allow me to display completed items (620ms)
      ✓ should allow me to display all items (2769ms)
      ✓ should highlight the currently applied filter (501ms)


  28 passing (4m)
```

**Comment by kmalakoff** [comment]
@samccone when I looked closer, I thought maybe the tests failed due to an install failure in the previous test.

```
> bufferutil@1.1.0 install /home/travis/build/tastejs/todomvc/tests/node_modules/selenium-webdriver/node_modules/ws/node_modules/bufferutil
> node-gyp rebuild
gyp WARN install got an error, rolling back install
```

This most recent time, all of the installations passed but the tests still failed.

```
TodoMVC - knockback
    When page is initially opened
  ✓ should focus on the todo input field
    No Todos
  1) should hide #main and #footer
    New Todo
  2) should allow me to add todo items
  ✓ should clear text input field when an item is added (199ms)
  3) should append new items to the bottom of the list
  4) should trim text input
  ✓ should show #main and #footer when items added (309ms)
    Mark all as completed
  5) should allow me to mark all items as completed
  6) should allow me to clear the completion state of all items
  7) complete all checkbox should update state when items are completed / cleared
    Item
  8) should allow me to mark items as complete
  9) should allow me to un-mark items as complete
    Editing
  10) "before each" hook for "should hide other controls when editing"
    Counter
  11) should display the current number of todo items
    Clear completed button
  12) should display the correct text
  13) should remove completed items when clicked
  14) should be hidden when there are no items that are completed
    Persistence
  15) should persist its data
    Routing
  16) should allow me to display active items
  17) should respect the back button
  18) should allow me to display completed items
  19) should allow me to display all items
  20) should highlight the currently applied filter
```

Is there a known timing problem with the test runner? Knockback listens for the DOM to be ready before injecting its view models and the local store should be synchronous, but it looks like selenium might be checking for something before an operation has taken place: 

```
  10) TodoMVC - knockback Editing "before each" hook for "should hide other controls when editing":
     NoSuchElementError: no such element: Unable to locate element: {"method":"xpath","selector":"//ul[@class="todo-list"]/li[2]//label"}
  (Session info: chrome=47.0.2507.0)
  (Driver info: chromedriver=2.19.346067 (6abd8652f8bc7a1d825962003ac88ec6a37a82f1),platform=Linux 3.13.0-40-generic x86_64)
      at new bot.Error (node_modules/selenium-webdriver/lib/atoms/error.js:108:18)
      at Object.bot.response.checkResponse (node_modules/selenium-webdriver/lib/atoms/response.js:109:9)
      at node_modules/selenium-webdriver/lib/webdriver/webdriver.js:379:20
      at promise.ControlFlow.runInFrame_ (node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:1857:20)
      at goog.defineClass.notify (node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:2448:25)
      at promise.Promise.notify_ (node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:564:12)
      at Array.forEach (native)
      at promise.Promise.notifyAll_ (node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:553:15)
      at goog.async.run.processWorkQueue (node_modules/selenium-webdriver/lib/goog/async/run.js:125:21)
  From: Task: TodoMVC - knockback Editing "before each" hook
```

Any thoughts? Any ideas how to troubleshoot?

**Comment by samccone** [comment]
Hey @kmalakoff sorry the delay.
**TLDR; things working locally and not on CI is no so surprising.**

To start off take a look :eyes: at this issue https://github.com/tastejs/todomvc/pull/1371 :book: 



In your case it looks like you are hitting an aync race condition between when DOM nodes are inserted and when we try and interact with them. thus :boom: . 

#### The fix :hammer: 

Assume CI is the slowest computer in the world got it? Ok now put that computer on the edge of an event horizon. Now you have an approximation for how slow CI is.

To debug / fix the failing specs in your app, I would use the pattern outlined above of checking to see if an element is present before trying to interact with it. Hopfully that should resolve your issue and also harden our test suite against further async issues of the nature. :clap: 

I **really** appreciate the work you are putting into this :a: :heavy_plus_sign: :clap: :palm_tree: :sun_with_face:

**Comment by kmalakoff** [comment]
@samccone Wow! There is a lot to process in there.

If I understand correctly, it sounds like there is not much to be done on my side...

1) Is the conclusion that the test suite needs to be made more robust? 
2) Is this an existing problem for other frameworks or have they done something on their side to make the tests pass?

**Comment by samccone** [comment]
> 1) Is the conclusion that the test suite needs to be made more robust? 

yes

2) Is this an existing problem for other frameworks

For some... I typically just resolve the issues as I run into them

**Comment by kmalakoff** [comment]
Great! Can you resolve these ones please?  :wink:

I have already spent more time than expected on this submission and need to move on. Also, I'm using an old computer and the tests take a very long time to run so iteration is painfully slow. 

That said, if there are instructions for quickly and easily triggering the errors so I can iterate on some fixes locally, I could put in a bit more time to try to apply more of the fixes from https://github.com/tastejs/todomvc/pull/1371. If I cannot easily reproduce the problems locally, I'm not sure how to test them as being fixed except trial-and-error by keeping on updating my pull request. 

How do you fix them without going insane? :cyclone:

**Comment by kmalakoff** [comment]
@samccone maybe you can just waive the tests until they are fixed since they are passing locally?

**Comment by samccone** [comment]
hi @kmalakoff I was going to take a look at this, this weekend.

**Comment by kmalakoff** [comment]
That is really kind. I can only imagine the work required. Thank you!

At least, even more robust tests will benefit everyone.

My offer was genuine that it you can explain to me how to reproduce problems, I can dedicate some time to hardening the tests and learning selenium.

**Comment by samccone** [comment]
hi @kmalakoff looks like the failures are real :) you just forgot to vendor some files that you have locally gitignored :+1: the fact that the tests pass locally is actually a great sign :dancers: :dancers: :dancers: .... you only need to commit these files and we should be in business! 

screen shot 2015-09-11 at 7 33 32 pm

**Comment by kmalakoff** [comment]
Whoops! I flattened the node_modules in preparation for the flatness of npm3 and there was a copy-paste error in the .gitignore.

I'm really sorry to have taken up your time on this. Thank you for tracking this down and supporting TodoMVC...awesome work!

All passing and ready to merge!

**Comment by kmalakoff** [comment]
@samccone just wondering if this pull request can be merged in? 

Let me know if there is anything else I need to do!

**Comment by samccone** [comment]
oh word just need to :eyes: the code once more :)

**Comment by kmalakoff** [comment]
Okee dokee. Let me know if you see anything.

**Comment by samccone** [comment]
Can you squash the following commits together


66b0657ab3d2c45c61565b332d87f9828f948edc
075eb72059ae5ddfd6477ec62678bde1510da766
de8e96f91c9f9279492c2412bd7985592c78a9b1


and then

60d4ffdfc1984a274e9b18d52668f02be7f5fa19
fd127ade5ea11a2e88dec52bf4e857b7eca830d9


:+1:

**Comment by kmalakoff** [comment]
I'm not sure how to do that. I'm not really a big git expert.

I could rollback all of my changes, apply them on top of head and open a new pull request.

**Comment by samccone** [comment]
@kmalakoff let me take a stab real quick

**Comment by kmalakoff** [comment]
Great! Thank you

**Comment by samccone** [comment]
all landed! :airplane: :baggage_claim: :walking: :handbag: :taxi: :beers: :cocktail: :tropical_drink: :hotel: :sleeping: 

Thanks again @kmalakoff :a: :heavy_plus_sign: work here

**Comment by kmalakoff** [comment]
Right on! Thanks for all the support including the git magic! 

:+1: :heavy_multiplication_x: a bazillion
